src/modules/flow/oauth/oauth.c

```diff
@@ -0,0 +1,596 @@
+            if (streq(value->value.key_value.key, "oauth_verifier")) {
+                verifier = strdup(value->value.key_value.value);
+            }
+            if (streq(value->value.key_value.key, "oauth_token")) {
+                token = strdup(value->value.key_value.value);
+            }
+            break;
+        default:
+            break;
+        }
+    }
+
+    SOL_NULL_CHECK_GOTO(verifier, err);
+
+    req_data->request = request;
+    req_data->mdata = data;
+    if (!sol_http_param_add(&params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD("oauth_token", token)) ||
+        !sol_http_param_add(&params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD("oauth_verifier", verifier))) {
+        goto err_params;
+    }
+
+    connection = sol_http_client_request(SOL_HTTP_METHOD_POST, mdata->access_token_url,
+        &params, v1_access_finished, req_data);
+    SOL_NULL_CHECK_GOTO(connection, err_connection);
+
+    r = sol_ptr_vector_append(&mdata->pending_conns, connection);
+    SOL_INT_CHECK_GOTO(r, < 0, err_append);
+
+    sol_http_param_free(&params);
+    free(verifier);
+    free(token);
+    return 0;
+
+err_append:
+    sol_http_client_connection_cancel(connection);
+err_connection:
+err_params:
+    sol_http_param_free(&params);
+err:
+    free(req_data->timestamp);
+    free(req_data->nonce);
+    free(req_data);
+    free(token);
+    free(verifier);
+    r = sol_http_server_send_response(request, &((struct sol_http_response) {
+            .api_version = SOL_HTTP_RESPONSE_API_VERSION,
+            .content = SOL_BUFFER_INIT_EMPTY,
+            .param = SOL_HTTP_REQUEST_PARAM_INIT,
+            .response_code = SOL_HTTP_STATUS_INTERNAL_SERVER_ERROR
+        }));
+    return r;
+}
+
+static void
+v1_close(struct sol_flow_node *node, void *data)
+{
+    struct sol_http_client_connection *connection;
+    struct v1_data *mdata = data;
+    uint16_t i;
+
+    free(mdata->request_token_url);
+    free(mdata->authorize_token_url);
+    free(mdata->access_token_url);
+    free(mdata->namespace);
+    free(mdata->consumer_key);
+    free(mdata->consumer_key_secret);
+    free(mdata->callback_url);
+    SOL_PTR_VECTOR_FOREACH_IDX (&mdata->pending_conns, connection, i)
+        sol_http_client_connection_cancel(connection);
+    sol_ptr_vector_clear(&mdata->pending_conns);
+}
+
+static int
+v1_parse_response(struct v1_request_data *req_data, const struct sol_http_response *response)
+{
+    int r = -EINVAL;
+    char *token, *token_secret, *url = NULL;
+    struct sol_http_response start_response = {
+        .api_version = SOL_HTTP_RESPONSE_API_VERSION,
+        .content = SOL_BUFFER_INIT_EMPTY,
+        .param = SOL_HTTP_REQUEST_PARAM_INIT,
+        .response_code = SOL_HTTP_STATUS_FOUND,
+    };
+
+    token = strstr(response->content.data, "oauth_token=");
+    token_secret = strstr(response->content.data, "oauth_token_secret=");
+
+    if (token_secret) {
+        token = strndup(token, token_secret - token - 1);
```

Comment:
This won't work if `oauth_token=` and `oauth_token_secret=` are in a different order.